lib/kino/js/live.ex

```diff
@@ -0,0 +1,369 @@
+  `main.js` file and define the `new(html)` function for
+  building the widget. As a matter of fact, the initial result
+  of `Kino.LiveHTML.new(html)` will render exactly the same
+  as our previous `Kino.HTMl.new(html)`.
+
+  As for the new bits, we added `use Kino.JS.Live` to define
+  a live widget server. We use `Kino.JS.Live.new/2` for creating
+  the widget instance and we implement a few `GenServer`-like
+  callbacks.
+
+  Once the widget server is started with `Kino.JS.Live.new/2`,
+  the `c:init/2` callback is called with the initial argument.
+  In this case we store the given `html` in server state.
+
+  Whenever the widget is rendered on the client, the `c:handle_connect/1`
+  callback is called and it builds the initial data for the
+  client. In this case, we always return the stored `html`.
+  This initial data is then passed to the JavaScript `init`
+  function. Keep in mind that while the server is initialized
+  once, connect may happen at any point, as the users join/refresh
+  the page.
+
+  Finally, the whole point of our example is the ability to
+  replace the HTML content directly from the Elixir side and
+  for this purpose we added the public `replace(widget, html)`
+  function. Underneath the function uses `cast/2` to message
+  our server and the message is handled with `c:handle_cast/2`.
+  In this case we store the new `html` in the server state and
+  broadcast an event with the new value. On the client side,
+  we subscribe to those events with `ctx.handleEvent(event, callback)`
+  to update the page accordingly.
+
+  ## Event handlers
+
+  You must eventually register JavaScript handlers for all events
+  that the client may receive. However, the registration can be
+  deferred, if the initialization is asynchronous. For example,
+  the following is perfectly fine:
+
+  ```js
+  export function init(ctx, data) {
+    fetch(data.someUrl).then((resp) => {
+      ctx.handleEvent("update", (payload) => {
+        // ...
+      });
+    });
+  }
+  ```
+
+  Or alternatively:
+
+  ```js
+  export async function init(ctx, data) {
+    const response = await fetch(data.someUrl);
+
+    ctx.handleEvent("update", (payload) => {
+      // ...
+    });
+  }
+  ```
+
+  In such case all incoming events are buffered and dispatched once
+  the handler is registered.
+  '''
+
+  defstruct [:module, :pid]
+
+  @doc false
+  use GenServer
+
+  require Logger
+
+  alias Kino.JS.Live.Context
+
+  @type t :: %__MODULE__{module: module(), pid: pid()}
+
+  @doc """
+  Invoked when the widget server started.
+
+  See `c:GenServer.init/1` for more details.
+  """
+  @callback init(arg :: term(), ctx :: Context.t()) :: {:ok, ctx :: Context.t()}
+
+  @doc """
+  Invoked whenever a new client connects to the server.
+
+  The returned data is passed to the JavaScript `init` function
+  of the connecting client.
+  """
+  @callback handle_connect(ctx :: Context.t()) :: {:ok, data :: term(), ctx :: Context.t()}
```

Comment:
I'm thinking we should probably make `origin` available in `handle_connect` and `handle_event` to have a way of distinguishing clients. It could be an argument or a map with metadata (in case we want to add more info in the future). Thoughts?

Reply:
Map with metadata sounds good to me.

Reply:
Another option is to have this information in the context, so you don't need to pass extra arguments on `handle_connect` and `handle_event`.

Reply:
Interesting idea, my only concern would be that conceptually `ctx` is information about the server state, which is shared across all clients. But then it's already called context so it can be anything :see_no_evil:

Reply:
Any way is fine by me. Just adding options. IMO it is fine for `ctx` to have the information about what triggered the callback. :D

Reply:
Ah, one thing is that for callbacks other than `handle_connect` and `handle_event` it would be `nil`, but that's probably fine. In the end it's only useful in specific cases, so I like not needing to match-ignore it most of the time :)

Reply:
Added `ctx.origin` in 147be9334fe06ca065c940b8f301c870ee4ece15.